nucypher-core/src/versioning.rs

```diff
@@ -136,6 +136,30 @@ pub trait ProtocolObjectInner<'a>: Serialize + Deserialize<'a> {
     fn unversioned_to_bytes(&self) -> Box<[u8]>;
 
     fn unversioned_from_bytes(minor_version: u16, bytes: &'a [u8]) -> Option<Result<Self, String>>;
+
+    /// Returns whether backwards incompatible changes (major version bumps) are allowed for
+    /// the object, and the frozen major version for which compatibility is frozen. If backwards
+    /// incompatible changes are allowed, simply return the current major version (default).
+    ///
+    /// There are legacy protocol object types that have already made major version changes
+    /// that should have remained backwards compatible; it's possible that it was versioned
+    /// during development but earlier versions were never released. This is the ONLY reason
+    /// why frozen major version is returned, but in most cases this value will be "1" i.e. the
+    /// first iteration of the type.
+    ///
+    /// Examples of where this is needed is for protocol objects that may be persisted
+    /// and then processed by later versions of the `nucypher-core` library eg. MessageKits
+    /// i.e. encrypted data. These protocol types need to be "perpetually" backwards compatible.
+    /// If any backwards incompatible changes are needed for types that are supposed to
+    /// always remain backwards compatible, then use a brand new type.
+    ///
+    /// NOTE: Underlying struct members for types that must remain backwards compatible should
+    /// also remain backwards compatible as well.
+    fn must_remain_backwards_compatible() -> (bool, u16) {
```

Comment:
I understand that is the solution for smooth switching from development to the production version?

Not sure how I feel about it. Seems kind of brittle. Maybe just a different type would be the way to go. You'd have to do some additional work when switching to production, but that would solidify the commitment, so to speak. The new type may still save/check the last development major version in the serialized form, to allow the usage of objects created during development, but I'm not sure how much the convenience here outweighs the safety. Perhaps it is better to force the devs to start the production objects from a clean slate.

Reply:
Let's muse more about this in https://github.com/nucypher/nucypher-core/pull/75.